script/translate.php

```diff
@@ -0,0 +1,17 @@
+<?php
+
+use SilverStripe\TxTranslator\Translator;
+use Symfony\Component\Dotenv\Dotenv;
+
+$vendorDir = dirname(dirname(dirname(__DIR__)));
+require "$vendorDir/autoload.php";
```

Comment:
```suggestion
$candidates = [
    // module in vendor
    __DIR__ . '/../../../autoload.php',
    // module itself is root (in its own dir instead of as a dependency)
    __DIR__ . '/../vendor/autoload.php',
];
$required = false;
foreach ($candidates as $candidate) {
    if (file_exists($candidate ?? '')) {
        require_once $candidate;
        $required = true;
    }
}
if (!$required) {
    die("Failed to include composer's autoloader, unable to continue");
}
```
This will allow us to run it from its own directory if we want, instead of as a dependency in a project.
Note I grabbed this snipped form framework which is why the `dirname` was replaced with `..` - feel free to use the `dirname` approach instead if you prefer.

Reply:
It won't work like that without modifying this to also work as a non-vendor-module.

There's code in `setModulePaths()` which assumes this is installed as a vendor module `$vendorDir = dirname(dirname(dirname(__DIR__))); ... `foreach (scandir($vendorDir) as $vendor) {` ... ` foreach (scandir("$vendorDir/$vendor") as $module) {`

`setModulePaths()` is used to know where to look for the translation files to update

We'd need to add extra code to pass in TX_PROJECT_PATH=/path/to/site-to-scan ... which I really don't see why we should spend any time to get that to work.  That's also an extra way to introduce human error e.g. you type in a path to the wrong site

Seems much easier to just say you have to composer install it on the project you intend to use to do the updates

Reply:
> There's code in `setModulePaths()` which assumes this is installed as a vendor module `$vendorDir = dirname(dirname(dirname(__DIR__)));` ... `foreach (scandir($vendorDir) as $vendor) {`... `foreach (scandir("$vendorDir/$vendor") as $module) {`

I had assumed you'd do that from the perspective of where it's run, given your instructions already assume you're at the project root for the site (your path is hardcoded in your instructions as `vendor/silverstripe/tx-translator/script/translate.php`). With that perspective, the above in `setModulePaths()` simply changes to `$vendorDir = getcwd() . '/vendor';` That also means we can move that logic around into a different file path in the future (maybe we eventually find we want `src/SomeNamespace/Translator.php` for whatever reason) and it'll still work because it's not based on the path of the file.

> We'd need to add extra code to pass in TX_PROJECT_PATH=/path/to/site-to-scan

No we wouldn't. Just assume you're running from the project root - which your readme instructions _already assume_. Make it explicit that this is the case, and if you want to even run a validation `if (!is_dir(getcwd() . '/vendor/silverstripe/framework')) {throw new Exception('Run this from the project root');}`

> By support for instance docs like `TX_SITE=mylocalsite.test php vendor/silverstripe/tx-translator/script/translate.php` .. now we need to specify and test 2x different ways to do this, which seems confusing for such a simple thing.

This is a solved "problem" - my recommendation in the readme does this very clearly already, and the current instruction is currently making assumptions about how you'll run it without being explicit which will lead to more confusion than explicitly stating "here are the two ways you can do this".

In any case, being able to run from an arbitrary directory isn't two ways to run it. Whether you `git clone` or `composer require` this thing, you're still going to have _the same code_ in some directory, and you'll navigate to the project root of the site you want to run it against, and you'll run `TX_SITE=my-local-site.test TX_PUSH=1 php path/to/tx-translator/script/translate.php` whether that path is `vendor/silverstripe/tx-translator/script/translate.php` or something else, you're still doing the exact same thing.

> That's also an extra way to introduce human error e.g. you type in a path to the wrong site

That's already possible if you have this module installed in multiple sites. You can have a "You are running this in [folder name] which has [xyz silverstripe dependency constraints] - is that what you expected? y/n" or "This will create PRs for the [framework PR branch] compatible branches - is that what you expected? y/n" prompt if you're really worried about that but I think that's not any more likely with my recommendation than the way it currently is.


## TL;DR
This needs _really minimal changes_ to work from some arbitrary directory, and I think it's a good thing to let it be flexible to run that way.

Reply:
> Just assume you're running from the project root 

Yeah, I am assuming that.  I don't understand why we need to add any code given that's the assumption?

> This needs really minimal changes to work from some arbitrary directory

This contradicts what you just said saying to assume we're running it from the project root?

I don't see any reason to not run this from a non-project root?

I simply can't understand what you're asking for

Reply:
When I said "running from the project root" I mean if you go `pwd` it will show you you're in the project root.
But the directory holding this code could be anywhere.

Reply:
> I simply can't understand what you're asking for

I'm asking for this to stop assuming it's installed as a module, and instead just check for the vendor dir in `pwd` for anything it needs re: translation and check for the autoloading in a way similar to my suggested change.
That's _all_ you need to change to let this be flexible. And yes it's only four (not three) of us using it at the moment. But the current restriction which is presumed and not even explicitly stated in the readme seems unnecessary when just these two simple changes allow us to run it how we want.

Reply:
Put it this way:
There's no downside to letting us run this in a more flexible way beyond the _very minimal_ maintenance increase presented by the new autoloading logic, and _potentially_ a slightly increased chance of running it against the wrong project, which is a problem that already potentially exists and I've already suggested ways to mitigate that.

Reply:
This module is installed via composer on an existing project

Then you run the script

I don't understand why we need any other setup than that

Reply:
I... have explained why that would be useful? Several times? And even if you don't understand why to do it, I've clearly got reasons why I think it's a good idea (that I've shared) and there doesn't seem to be any meaningful downside, and it's easy to do, so... why not just do it?
But honestly if you're going to just not understand and therefore not do it then there's no amount of explaining or pleading I can do to get this change in, so I'll just resolve the conversation.